server/src/testFixtures/java/io/crate/testing/SQLTransportExecutor.java

```diff
@@ -153,11 +159,40 @@ public long[] execBulk(String statement, @Nullable Object[][] bulkArgs, TimeValu
         return executeBulk(statement, bulkArgs, timeout);
     }
 
+    @VisibleForTesting
+    static List<String> buildRandomizedRuleSessionSettings(Random random,
+                                                           double factor,
+                                                           List<Class<? extends Rule<?>>> allRules,
+                                                           List<Class<? extends Rule<?>>> rulesToKeep) {
+        assert factor >= 0 && factor <= 1 : "Factor for Rule Randomization must between 0 and 1";
+
+        var ruleToKeepNames = new HashSet<>(rulesToKeep);
+
+        var ruleCandidates = new ArrayList<Class<? extends Rule<?>>>();
+        for (var rule : allRules) {
+            if (ruleToKeepNames.contains(rule) == false) {
+                ruleCandidates.add(rule);
+            }
+        }
+
+        Collections.shuffle(ruleCandidates, random);
+        int numberOfRulesToPick = (int) Math.ceil(ruleCandidates.size() * factor);
```

Comment:
I understand what is happening here but there are some edge cases. For example, if `ruleCandiates.size()` is `1` and `factor` is `0.1` I would expect `0.1` chance of the single rule applied but in this case it will always be applied.
But the effect is pretty minor, the worst thing that can happen is to test a few more (or less) rules time to time.
LGTM otherwise. 👍

Reply:
Good catch, thank you. I had originally originally:
```
int numberOfRulesToPick = random.nextInt(1, ruleToKeepNames.size());
```
this would have avoided that but made it tricky to test. I will do another iteration.

Reply:
In practice `ruleCandiates.size()` will always be more than 10, but i think this should work correctly for all cases.

Reply:
What is the drawback if we don't use `ceil()` but rounding?
if @jeeminso 's example with 1 rule. if factor >=0.5 then 1 rule if factor < 0.5 then 0 rules.

Reply:
To be precise, the edge case still exists even if we use `round()` here. If `ruleCandidates.size() = 2` and `factor = 0.5` then `numberOfRulesToPick = 1` always. But in reality we sometimes expect `numberOfRulesToPick = 2` or `numberOfRulesToPick = 0`.

Reply:
Apologies if I miss something, the factor is random double between `0.0` and `1.0`, so why do you say always?

Reply:
You are right, I must have assumed that `UseRandomizedOptimizerRules.value` would be fed in as the `factor`.

Reply:
The factor is the percentage of rules which should be deactivated, not the chances that rules are deactivated (randomized). See https://github.com/crate/crate/pull/13851#discussion_r1147282310
So with the current approach:
```
numberOfRulesToPick = (int) Math.ceil(ruleCandidates.size() * factor);
``` 
we are rounding up which would bring the following outcomes to the edge cases:
`1 Rule, 0.1 Factor -> 1 Rule is deactivated`
`2 Rules, 0.5 Factor -> 1 Rule is deactivated`
which I think is acceptable. If we change to:
```
numberOfRulesToPick = Math.round(ruleCandidates.size() * factor);
``` 
`1 Rule, 0.1 Factor -> 0 Rule is deactivated`
`2 Rules, 0.5 Factor -> 1 Rule is deactivated`
which i think is less expected in the first example. Because we assume that randomization is happening but nothing is happening at all.

In general, there will be always the full ruleset of candidates to choose from which is more than 20.

Reply:
I agree, thx for explaining! ceiling is a better choice.

Reply:
👍